File: server/Services/CryptidEncountersService.cs




namespace cryptipedia.Services;

public class CryptidEncountersService
{
  public CryptidEncountersService(CryptidEncountersRepository repository)
  {
    _repository = repository;
  }
  private readonly CryptidEncountersRepository _repository;

  internal CryptidEncounterProfile CreateCryptidEncounter(CryptidEncounter cryptidEncounterData)
  {
    CryptidEncounterProfile cryptidEncounterProfile = _repository.CreateCryptidEncounter(cryptidEncounterData);
    return cryptidEncounterProfile;
  }

  internal List<CryptidEncounterProfile> GetCryptidEncounterProfilesByCryptidId(int cryptidId)
  {
    List<CryptidEncounterProfile> cryptidEncounterProfiles = _repository.GetCryptidEncounterProfilesByCryptidId(cryptidId);
    return cryptidEncounterProfiles;
  }

  internal List<CryptidEncounterCryptid> GetCryptidEncountersByAccountId(string accountId)
  {
    List<CryptidEncounterCryptid> cryptids = _repository.GetCryptidEncountersByAccountId(accountId);
    return cryptids;
  }

  internal void DeleteCryptidEncounter(int cryptidEncounterId, Account userInfo)
  {
    CryptidEncounter cryptidEncounter = GetCryptidEncounterById(cryptidEncounterId);

    if (cryptidEncounter.AccountId != userInfo.Id)
    {
      throw new Exception($"YOU CANNOT DELETE ANOTHER USER'S ENCOUNTER, {userInfo.Name.ToUpper()}!!!");
    }

    _repository.DeleteCryptidEncounter(cryptidEncounterId);
  }

  private CryptidEncounter GetCryptidEncounterById(int cryptidEncounterId)
  {
    CryptidEncounter cryptidEncounter = _repository.GetCryptidEncounterById(cryptidEncounterId);

    if (cryptidEncounter == null)
    {
      throw new Exception("Invalid id: " + cryptidEncounterId);
    }

    return cryptidEncounter;
  }
}

File: client/src/services/CryptidEncountersService.js
import { logger } from "@/utils/Logger.js"
import { api } from "./AxiosService.js"
import { CryptidEncounterProfile } from "@/models/CryptidEncounterProfile.js"
import { AppState } from "@/AppState.js"
import { CryptidEncounterCryptid } from "@/models/CryptidEncounterCryptid.js"

class CryptidEncountersService {
  async getCryptidEncounterProfilesByCryptidId(cryptidId) {
    AppState.cryptidEncounterProfiles.length = 0
    const response = await api.get(`api/cryptids/${cryptidId}/cryptidEncounters`)
    logger.log('GOT CRYPTID ENCOUNTERS', response.data)
    const profiles = response.data.map(pojo => new CryptidEncounterProfile(pojo))
    AppState.cryptidEncounterProfiles = profiles
  }

  async getMyCryptidEncounters() {
    const response = await api.get('account/cryptidEncounters')
    logger.log('GOT MY CRYPTID ENCOUNTERS', response.data)
    const cryptids = response.data.map(pojo => new CryptidEncounterCryptid(pojo))
    AppState.cryptidEncounterCryptids = cryptids
  }

  async createCryptidEncounter(cryptidEncounterData) {
    const response = await api.post('api/cryptidEncounters', cryptidEncounterData)
    logger.log('ENCOUNTERED CRYPTID', response.data)
    const profile = new CryptidEncounterProfile(response.data)
    AppState.cryptidEncounterProfiles.push(profile)
  }


  async deleteCryptidEncounter(cryptidEncounterId) {
    const response = await api.delete(`api/cryptidEncounters/${cryptidEncounterId}`)
    logger.log('DELETED!', response.data)
    const cryptidEncounters = AppState.cryptidEncounterCryptids
    const index = cryptidEncounters.findIndex(cryptid => cryptid.cryptidEncounterId == cryptidEncounterId)

    const cryptidEncounter = cryptidEncounters[index]

    cryptidEncounters.forEach(cryptid => {
      if (cryptid.id != cryptidEncounter.id) return
      cryptid.encounterCount--
    });

    cryptidEncounters.splice(index, 1)
  }

}

export const cryptidEncountersService = new CryptidEncountersService()